### DISCUSSION TITLE
Enable and disable fields with an "Edit" button

### DISCUSSION THREAD

**Discussion by acseven** [Budibase/budibase#9536] [Help] [answered]
I have a form view that I would like to be opened with all its fields locked / disabled / read only and instead of duplicating the form to have a second one to edit the fields, is it possible to use a button action to enable the fields, and disable them again on a "cancel edit" or "save" button?

**Comment by melohagan** [comment, marked as answer]
Hi @acseven 

Interesting use-case! Here's an approach you could take.

1. Create an app with the sample data

2. Add a screen called `Inventory`. Add a table block for the Inventory table
![Screenshot 2023-02-13 at 14 59 18](https://user-images.githubusercontent.com/101575380/218492759-6f9e6285-958c-4eb9-a3fc-e7b7a614e0cf.png)

3. `Define actions` for the On Row Click as follows:
![Screenshot 2023-02-13 at 15 02 01](https://user-images.githubusercontent.com/101575380/218493541-e8f8b38c-7535-4776-a731-a41dc9cdb0b1.png)

4. Create another screen with the route `/inventory/:id`. Add an *Update* form block for the Inventory table with the *Row ID* set to `{{ URL.id }}`. De-select. any unwanted fields.
![Screenshot 2023-02-13 at 15 04 42](https://user-images.githubusercontent.com/101575380/218494227-b17340a4-7731-40f5-9300-500fe9121e49.png)

5. In the *On screen load* actions of the `/inventory/:id` screen, `Define actions` to set an initial lock state as follows:
![Screenshot 2023-02-13 at 15 08 26](https://user-images.githubusercontent.com/101575380/218495066-e34b147b-a29f-4d11-9f25-c026c9f2b7fd.png)

With the JavaScript:
```js
return {
	itemName: true,
	itemTags: true,
	notes: true,
	status: true,
	purchaseDate: true,
	purchasePrice: true
}
```

6. This step will be a little tedious, but for each form field `Configure conditions` and check disabled if the lock state is true for that field. For example:
![Screenshot 2023-02-13 at 15 11 38](https://user-images.githubusercontent.com/101575380/218495836-b71ad6de-957a-4922-84af-424b754d71fb.png)
In this case the binding is: `{{ State.LockedFields.itemName }}`

7. Delete the `Save` button as we wont need it. Add three containers: 'Form Container', 'Buttons Container' and 'Item Name Buttons Container'. Drag the *Buttons Container* and the *Fieldgroup* into the *Form Container* and set its' layout to *Row Layout*. Set the *Item Name Buttons Container* to *Row Layout* as well.
![Screenshot 2023-02-13 at 15 48 17](https://user-images.githubusercontent.com/101575380/218505137-bf1c6c12-baf7-41ee-bf30-2419d2c7047c.png)

8. Add three buttons to the *Item Name Buttons Container*: Edit, Cancel, Save. `Configure Conditions` for each of the buttons (feel free to duplicate the buttons for convenience). The *Edit* button condition should be as follows:
![Screenshot 2023-02-13 at 16 10 54](https://user-images.githubusercontent.com/101575380/218510916-01715896-9394-4b5a-86b6-6fe41b2484b6.png)
Where the binding is: `{{ State.LockedFields.itemName }}`. 

Do the same for the *Cancel* and *Save* buttons, however the boolean flag would be **True** in those cases.

9. Add an **Update State** action for the *Edit* button like so:
![Screenshot 2023-02-13 at 16 13 10](https://user-images.githubusercontent.com/101575380/218511485-92aceaba-9049-4865-ace8-a49cc3dae9eb.png)

With the JavaScript binding:
```js
return {
	...$("State.LockedFields"),
	itemName: false
}
```

Do the same for *Cancel* and *Save*, but flip the boolean to **true**. For the *Save* button add another action:
![Screenshot 2023-02-13 at 16 18 17](https://user-images.githubusercontent.com/101575380/218512644-67fe857e-0b46-4374-86d5-d38781a6c8c5.png)

And for the *Cancel* button, add another action as follows to reset the field value:
![Screenshot 2023-02-13 at 16 17 09](https://user-images.githubusercontent.com/101575380/218512357-aa2affb4-6cbc-4503-8960-26c73eb72e77.png)

10. Duplicate the *Item Name Buttons Container* for each field and update the state references to the appropriate property. Add margin top to the buttons containers to space the buttons as desired.
![Screenshot 2023-02-13 at 16 28 35](https://user-images.githubusercontent.com/101575380/218515064-7922baae-a802-41b8-ad49-2ed24ce0b305.png)

Also give the *Buttons Container* a fixed width so that the form fields don't move when switching between one Edit button, and two Cancel/Save buttons, say `250px`.

### Result
![inline_edit](https://user-images.githubusercontent.com/101575380/218515859-04c2c2df-7102-443b-83f1-91936046b422.gif)

#### App Export
[Edit field form-export-1676305947437.tar.gz](https://github.com/Budibase/budibase/files/10724211/Edit.field.form-export-1676305947437.tar.gz)

**Comment by acseven** [comment]
Thanks a bunch for the detailed instructions @melohagan! I've replicated it in a more general approach (a single Edit/Cancel/Save buttons group) and it seems to work well! 
For large forms it's somewhat of an hassle to setup all the fields and especially hard to maintain given there are so many changes in several different places, but nonetheless works great!
I hope one day this type of editing could be easier to implement out-of-the-box!

Thank you again!

**Reply by t1mmyn4t0r** [reply]
Hi @melohagan , 
I also need this in my app, but instead of editing each field separately, do you have a way I can click 1 edit button and edit 3 to 5 fields and then save them when done?

**Reply by melohagan** [reply]
I think that case would be easier! [Forms](https://docs.budibase.com/docs/form-block) have different modes: 

- Create
- Edit
- View

You could have two forms, one for edit and another for read-only view, and [hide/show component](https://docs.budibase.com/docs/conditions) depending on how a button is toggled, which would require [app state](https://docs.budibase.com/docs/app-state) to keep track of a boolean flag.

**Reply by t1mmyn4t0r** [reply]
Thank you for the fast reply!
I am a noob to programming, i started with appsheet and decided to move to budibase for more variety. What code should i enter where?
